enterprise/wsproxy/proxy.go

```diff
@@ -0,0 +1,251 @@
+// directly with a workspace. It requires a primary coderd to establish a said
+// connection.
+type Server struct {
+	Options *Options
+	Handler chi.Router
+
+	PrimaryAccessURL *url.URL
+	AppServer        *workspaceapps.Server
+
+	// Logging/Metrics
+	Logger             slog.Logger
+	TracerProvider     trace.TracerProvider
+	PrometheusRegistry *prometheus.Registry
+
+	// SDKClient is a client to the primary coderd instance authenticated with
+	// the moon's token.
+	SDKClient *wsproxysdk.Client
+
+	// TODO: Missing:
+	//		- derpserver
+
+	// Used for graceful shutdown. Required for the dialer.
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+func New(opts *Options) (*Server, error) {
+	if opts.PrometheusRegistry == nil {
+		opts.PrometheusRegistry = prometheus.NewRegistry()
+	}
+
+	if err := opts.Validate(); err != nil {
+		return nil, err
+	}
+
+	// TODO: implement some ping and registration logic
+	client := wsproxysdk.New(opts.PrimaryAccessURL)
+	err := client.SetSessionToken(opts.ProxySessionToken)
+	if err != nil {
+		return nil, xerrors.Errorf("set client token: %w", err)
+	}
+
+	r := chi.NewRouter()
+	ctx, cancel := context.WithCancel(context.Background())
+	s := &Server{
+		Options:            opts,
+		Handler:            r,
+		PrimaryAccessURL:   opts.PrimaryAccessURL,
+		Logger:             opts.Logger.Named("workspace-proxy"),
+		TracerProvider:     opts.Tracing,
+		PrometheusRegistry: opts.PrometheusRegistry,
+		SDKClient:          client,
+		ctx:                ctx,
+		cancel:             cancel,
+	}
+
+	s.AppServer = &workspaceapps.Server{
+		Logger:        opts.Logger.Named("workspaceapps"),
+		DashboardURL:  opts.PrimaryAccessURL,
+		AccessURL:     opts.AccessURL,
+		Hostname:      opts.AppHostname,
+		HostnameRegex: opts.AppHostnameRegex,
+		RealIPConfig:  opts.RealIPConfig,
+		SignedTokenProvider: &ProxyTokenProvider{
+			DashboardURL: opts.PrimaryAccessURL,
+			AccessURL:    opts.AccessURL,
+			AppHostname:  opts.AppHostname,
+			Client:       client,
+			SecurityKey:  s.Options.AppSecurityKey,
+			Logger:       s.Logger.Named("proxy_token_provider"),
+		},
+		WorkspaceConnCache: wsconncache.New(s.DialWorkspaceAgent, 0),
+		AppSecurityKey:     opts.AppSecurityKey,
+
+		DisablePathApps:  opts.DisablePathApps,
+		SecureAuthCookie: opts.SecureAuthCookie,
+	}
+
+	// Routes
+	apiRateLimiter := httpmw.RateLimit(opts.APIRateLimit, time.Minute)
+	// Persistent middlewares to all routes
+	r.Use(
+		// TODO: @emyrk Should we standardize these in some other package?
+		httpmw.Recover(s.Logger),
+		tracing.StatusWriterMiddleware,
+		tracing.Middleware(s.TracerProvider),
+		httpmw.AttachRequestID,
+		httpmw.ExtractRealIP(s.Options.RealIPConfig),
+		httpmw.Logger(s.Logger),
+		httpmw.Prometheus(s.PrometheusRegistry),
+
+		// SubdomainAppMW is a middleware that handles all requests to the
+		// subdomain based workspace apps.
+		s.AppServer.SubdomainAppMW(apiRateLimiter),
```

Comment:
Because this is on the app server, this should be called `SubdomainMW` or `HandleSubdomain`.

Reply:
Or a more descriptive name would be helpful. There's a great comment on the middleware though, so it's fine if not.